RavenDB-6511 Smuggler for attachments

fitzchak opened the pull request.

Commit:
RavenDB-6511 Smuggler for attachments

ayende commented on src/Raven.Server/Documents/AttachmentsStorage.cs:
```diff
@@ -376,7 +377,7 @@ public DynamicJsonArray GetAttachmentsMetadataForDocument(DocumentsOperationCont
 
             var stream = GetAttachmentStream(context, attachment.Base64Hash);
-            if (stream == null)
+            if (stream == null || stream.Length == 0)
```
what happens if a user uploads a zero length file?

fitzchak replied:
Fixed, and I created a test that we support that.

ayende commented on src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:
```diff
@@ -137,12 +145,34 @@ public DatabaseDocumentActions(DocumentDatabase database, BuildVersionType build
+                HandleBatchOfDocumentsIfNecessary();
+            }
+
+            public void WriteAttachment(StreamSource.AttachmentStream attachment)
```
Are we allowed to trigger a batch because of an attachment?
If we do, we might write it before the actual document was written, no?

fitzchak replied:
This won't be an issue, but I removed the HandleBatchOfDocumentsIfNecessary from attachments.

ayende commented on src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs:
```diff
@@ -406,6 +411,36 @@ private SmugglerProgressBase.Counts ProcessDocuments(SmugglerResult result, Buil
+                return;
+
+            if (attachmentStreamsAlreadyExported == null)
+                attachmentStreamsAlreadyExported = new HashSet<LazyStringValue>();
```
Is it valid to save the LazyStringValue here? What context is it attached to, and are you sure it wouldn't be reset midway?

ayende commented on src/Raven.Server/Smuggler/Documents/StreamSource.cs:
```diff
@@ -344,6 +359,61 @@ private IEnumerable<Document> ReadDocuments(INewDocumentActions actions = null)
+        private unsafe void AddAttachmentStream(DocumentsOperationContext context, BlittableJsonReaderObject data, INewDocumentActions actions)
+        {
+            var documentActions = actions as DatabaseDestination.DatabaseDocumentActions;
+            if (documentActions == null)
```
Why, what else can this be, and why aren't we polymorphic here?

fitzchak replied:
Fixed.

Commit:
RavenDB-6511:
- Make sure to set the AttachmentResult.Size value.
- Add support for empty attachments. RavenCommand.ProcessResponse also always should dispose the Response.
- Add tests for export a document and change change its attachment (delete or add) and than import the old export.

Commit:
RavenDB-6511 Try to avoid NRE when the database is disposed. In that case dbInstance.IndexStore will be null. In debug I see that _token was not canceled but I do not sure how this is the case and if so maybe this fix is not enough.

Commit:
RavenDB-6511 There is no currently issue to finish a batch in the attachment without its document but we'll avoid this just to be sure we do not have cases like this over the all system.

ayende commented on src/Raven.Client/Documents/Operations/GetAttachmentOperation.cs:
```diff
@@ -135,6 +135,7 @@ public override void SetResponseUncached(HttpResponseMessage response, Stream st
                 var etag = response.GetRequiredEtagHeader();
                 IEnumerable<string> hashVal;
                 var hash = response.Headers.TryGetValues("Content-Hash", out hashVal) ? hashVal.First() : null;
+                var size = response.Content.Headers.ContentLength ?? 0;
```
What are you reading here, and what if the request is using `Chunked-Encoding` ?

fitzchak replied:
Thanks, fixed. 
Using stream.Length instead.

ayende commented on src/Raven.Server/Rachis/ClusterMaintenanceSlave.cs:
```diff
@@ -65,6 +65,9 @@ private IEnumerable<(string,DatabaseStatusReport)> CollectDatabaseInformation(Tr
         {
             foreach (var dbName in _server.Cluster.GetDatabaseNames(ctx))
```
@karmeli87 Please take a look
